src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,199 @@
+            tale = new Node(prevTale, value, null);
+            prevTale.next = tale;
+        } else if (size == 0) {
+            Node<T> fistNode = new Node<>(null, value, null);
+            head = fistNode;
+            tale = fistNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForBound(index, size);
+        if (index == size || index == size && size == 0) {
+            add(value);
+            return;
+        } else if (index == 0) {
+            Node<T> initialHeadNode = head;
+            head = new Node<>(null, value, initialHeadNode);
+            initialHeadNode.prev = head;
+            size++;
+            return;
+        }
+        if (index < (size - 1) / 2) {
+            int indexCounter = 0;
+            Node<T> tempNode = head.next;
+            while (indexCounter != index) {
+                indexCounter++;
+                if (indexCounter == index) {
+                    relinkNewNode(tempNode, value);
+                }
+                tempNode = tempNode.next;
+
+            }
+        } else {
+            int indexCounter = size;
+            Node<T> tempNode = tale;
+            while (indexCounter != index) {
+                indexCounter--;
+                if (indexCounter == index) {
+                    relinkNewNode(tempNode, value);
+                }
+                tempNode = tempNode.prev;
+            }
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexForBound(index, size - 1);
+        return getNode(index).body;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexForBound(index, size - 1);
+        if (index == 0) {
+            Node<T> initialHeadNode = head;
+            head = new Node<>(null, value, initialHeadNode.next);
+            initialHeadNode.next.prev = head;
+            return initialHeadNode.body;
+        } else if (index == size - 1) {
+            Node<T> initialTaleNode = tale;
+            tale = new Node<>(initialTaleNode.prev, value, null);
+            initialTaleNode.prev.next = tale;
+            return initialTaleNode.body;
+        } else {
+            Node<T> toRemoveNode = getNode(index);
+            Node<T> insteadNode = new Node<>(toRemoveNode.prev, value, toRemoveNode.next);
+            toRemoveNode.prev.next = insteadNode;
+            toRemoveNode.next.prev = insteadNode;
+            return toRemoveNode.body;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexForBound(index, size - 1);
+        Node<T> toRemoveNode = getNode(index);
+        if (index == 0) {
+            head = getNode(1);
+        } else if (index == size - 1) {
+            tale = getNode(size - 2);
```

Comment:
you don't need to iterate over nodes to find the head or tail. they are saved in list and can be accessed directly

Reply:
It's just reinitializing head or tail, if previous supposed to removed.